File: exercises/ad_mydual/ad.py
import math

class Dual:

    def __init__(self, value, deriv):
        self.value = value
        self.deriv = deriv

    def __str__(self):
        return "value: {0}\nderiv: {1}".format(self.value, self.deriv)

    def __add__(self, y):
        return Dual(self.value + y.value, self.deriv + y.deriv) 

    def __pow__(self, p):
        return Dual( , ) 

    def __truediv__(self, y):
        return Dual( , )

def sin(x):
    return Dual(math.sin(x.value), math.cos(x.value)*x.deriv)

def cos(x):
    return Dual( , )

def exp(x):
    return Dual( , )

def sqrt(x):
    return Dual( , )


def func(x):
    return exp(x)/sqrt(sin(x)**3 + cos(x)**3)

    
x = Dual(2.0, 1.0)
y = func(x)
print(y)
# exact derivative = 18.880984872278777

File: exercises/ad_mydual/ad.jl
struct Dual{TF}
    value::TF
    deriv::TF
end

Base.:sin(x::Dual) = Dual(sin(x.value), cos(x.value)*x.deriv)
Base.:cos(x::Dual) = Dual( , )
Base.:exp(x::Dual) = Dual( , )
Base.:sqrt(x::Dual) = Dual( , )
Base.:+(x::Dual, y::Dual) = Dual(x.value + y.value, x.deriv + y.deriv)
Base.:/(x::Dual, y::Dual) = Dual( , )
Base.:^(x::Dual, p::Number) = Dual( , )

func(x) = exp(x)/sqrt(sin(x)^3 + cos(x)^3)

x = Dual(2.0, 1.0)
y = func(x)
println(y)
# exact derivative = 18.880984872278777